compiler/codegen/mli.cpp

```diff
@@ -0,0 +1,912 @@
+  for (int i = 1; i <= fn->numFormals(); i++) {
+    ArgSymbol* as = fn->getFormal(i);
+    Type* t = getTypeFromFormal(as);
+
+    gen += this->genMarshalPushCall(client_arg, as->name, t);
+  }
+
+  // Pull and return result if applicable.
+  if (not hasVoidReturnType) {
+    gen += "result = ";
+    gen += this->genMarshalPullCall(client_res, "result", fn->retType);
+    gen += "return result;\n";
+  }
+ 
+  return gen;
+}
+
+std::string MLIContext::genServersideRPC(FnSymbol* fn) {
+  std::map<int, std::string> formalTempNames;
+  bool hasVoidReturnType = fn->retType == dtVoid;
+  bool hasFormals = fn->numFormals() != 0;
+  std::string gen;
+
+  // Emit void/void calls immediately, then return.
+  if (hasVoidReturnType and not hasFormals) {
+    gen += fn->cname;
+    gen += "();\n";
+    return gen;
+  }
+
+  // Declare a temporary for the return value, if necessary.
+  if (not hasVoidReturnType) {
+    gen += this->genTypeName(fn->retType);
+    gen += " result;\n";
+  }
+
+  // Declare temporaries, issue unpack call for each formal.
+  for (int i = 1; i <= fn->numFormals(); i++) {
+    std::string tmp;
+
+    Type* t = this->getTypeFromFormal(fn, i);
+
+    gen += this->genTypeName(t);
+    gen += " ";
+
+    // Map temp names to formal indices (shifted down one).
+    tmp += "tmp_";
+    tmp += std::to_string(i - 1);
+    formalTempNames[i] = tmp;
+
+    // Emit a unpack call to initialize each temporary.
+    gen += tmp;
+    gen += "=";
+    gen += this->genMarshalPullCall(server_arg, tmp.c_str(), t);
+  }
+
+  // Only generate LHS target if necessary.
+  if (not hasVoidReturnType) {
+    gen += "result=";
+  }
+
+  // Make the unwrapped call.
+  gen += fn->cname;
+  gen += "(";
+
+  // Pass in temporaries as arguments to call.
+  if (hasFormals) {
+    for (int i = 1; i <= fn->numFormals() - 1; i++) {
+      gen += formalTempNames.at(i);
+      gen += ",";
+    }
+    gen += formalTempNames.at(fn->numFormals());
+  }
+
+  gen += ");\n";
+
+  // If there is a result, issue a pack call for it.
+  if (not hasVoidReturnType) {
+    gen += this->genMarshalPushCall(server_res, "result", fn->retType);
+  }
+
+  return gen;
+}
+
+std::string MLIContext::genMarshalCall(const char* s, const char* v, Type* t,
+                                       bool out) {
+  std::string gen;
+  int64_t id = this->assignUniqueTypeID(t);
+
+  gen += out ? marshal_push_prefix : marshal_pull_prefix;
+  gen += std::to_string(id);
+  gen += "(";
+  gen += s;
+  
+  if (out) {
+    gen += ",";
+    gen += v;
+  }
+
+  gen += ");\n";
+
+  return gen;
+}
+
+//
+// TODO: These calls pass value types (cheaper to pass pointer).
+//
+std::string
+MLIContext::genMarshalPushCall(const char* s, const char* v, Type* t) {
+  return this->genMarshalCall(s, v, t, true);
+}
+
+//
+// TODO: These calls pass value types (cheaper to pass pointer).
+//
+std::string
+MLIContext::genMarshalPullCall(const char* s, const char* v, Type* t) {
+  (void) v;
```

Comment:
I'm confused by this line.  It looks like a no-op to me?

Reply:
Good catch. Holdover from when the type of that function was different.